shared/src/main/java/uk/gov/di/authentication/shared/services/DynamoAuthCodeService.java

```diff
@@ -0,0 +1,84 @@
+
+    public DynamoAuthCodeService(ConfigurationService configurationService) {
+        super(AuthCodeStore.class, "auth-code-store", configurationService);
+        this.timeToExist = configurationService.getAuthCodeExpiry();
+    }
+
+    public Optional<AuthCodeStore> getAuthCodeStore(String subjectID) {
+        return get(subjectID)
+                .filter(t -> t.getTimeToExist() > NowHelper.now().toInstant().getEpochSecond());
+    }
+
+    public void deleteAuthCode(String subjectID) {
+        delete(subjectID);
+    }
+
+    public void saveAuthCode(
+            String subjectID, String authCode, String requestedScopeClaims, boolean hasBeenUsed) {
+        var authCodeStore =
+                new AuthCodeStore()
+                        .withSubjectID(subjectID)
+                        .withAuthCode(authCode)
+                        .withRequestedScopeClaims(requestedScopeClaims)
+                        .withHasBeenUsed(hasBeenUsed)
+                        .withTimeToExist(
+                                NowHelper.nowPlus(timeToExist, ChronoUnit.SECONDS)
+                                        .toInstant()
+                                        .getEpochSecond());
+
+        put(authCodeStore);
+    }
+
+    public void updateHasBeenUsed(String subjectID, boolean hasBeenUsed) {
+        var authCode =
+                get(subjectID)
+                        .orElse(new AuthCodeStore())
+                        .withSubjectID(subjectID)
+                        .withHasBeenUsed(hasBeenUsed)
+                        .withTimeToExist(
+                                NowHelper.nowPlus(timeToExist, ChronoUnit.SECONDS)
+                                        .toInstant()
+                                        .getEpochSecond());
+
+        update(authCode);
+    }
+
+    public void updateAuthCode(String subjectID, String newAuthCode) {
+        var authCode =
+                get(subjectID)
+                        .orElse(new AuthCodeStore())
+                        .withSubjectID(subjectID)
+                        .withAuthCode(newAuthCode)
+                        .withTimeToExist(
+                                NowHelper.nowPlus(timeToExist, ChronoUnit.SECONDS)
+                                        .toInstant()
+                                        .getEpochSecond());
+
+        update(authCode);
+    }
+
+    public void updateRequestedScopeClaims(String subjectID, String requestedScopeClaims) {
+        var authCode =
+                get(subjectID)
+                        .orElse(new AuthCodeStore())
+                        .withSubjectID(subjectID)
+                        .withRequestedScopeClaims(requestedScopeClaims)
+                        .withTimeToExist(
+                                NowHelper.nowPlus(timeToExist, ChronoUnit.SECONDS)
+                                        .toInstant()
+                                        .getEpochSecond());
+
+        update(authCode);
```

Comment:
I reckon we won't need these methods. We will only ever create an auth code, update the has been used or read the auth code store